torchaudio/csrc/ray_tracing.cpp

```diff
@@ -0,0 +1,571 @@
+        scattering(_scattering),
+        normal(_normal),
+        origin(_origin) {}
+
+  torch::Tensor get_absorption() {
+    return absorption;
+  }
+  torch::Tensor get_reflection() {
+    return reflection;
+  }
+  torch::Tensor get_scattering() {
+    return scattering;
+  }
+
+  /**
+   * Returns the side (-1, 1 or 0) on which a point lies w.r.t. the wall.
+   */
+  int side(torch::Tensor pos) {
+    auto dot = VAL((pos - origin).dot(normal));
+
+    if (dot > EPS) {
+      return 1;
+    } else if (dot < -EPS) {
+      return -1;
+    } else {
+      return 0;
+    }
+  }
+
+  /**
+   * Reflects a ray (dir) on the wall. Preserves norm of vector.
+   */
+  torch::Tensor reflect(torch::Tensor dir) {
+    return dir - normal * 2 * dir.dot(normal);
+  }
+
+  /**
+   * Returns the cosine angle of a ray (dir) with the normal of the wall
+   */
+  scalar_t cosine(torch::Tensor dir) {
+    return VAL(dir.dot(normal) / dir.norm());
+  }
+
+ private:
+  torch::Tensor absorption;
+  torch::Tensor reflection; // == 1 - absorption
+  torch::Tensor scattering;
+  torch::Tensor normal; // The normal to the wall: 2D or 3D vector
+  torch::Tensor
+      origin; // The origin of the wall: corresponds to an arbitrary corner.
+};
+
+/**
+ * RayTracer class helper for ray tracing. For attribute description, please see
+ * declarations below as well as Python wrapper.
+ */
+template <typename scalar_t>
+class RayTracer {
+ public:
+  RayTracer(
+      const torch::Tensor _room,
+      const torch::Tensor _source,
+      const torch::Tensor _mic_array,
+      int _num_rays,
+      const torch::Tensor _e_absorption,
+      const torch::Tensor _scattering,
+      scalar_t _mic_radius,
+      scalar_t _sound_speed,
+      scalar_t _energy_thres,
+      scalar_t _time_thres,
+      scalar_t _hist_bin_size)
+      : room(_room),
+        source(_source),
+        mic_array(_mic_array),
+        num_rays(_num_rays),
+        energy_0(2. / num_rays),
+        // e_absorption(_e_absorption),
+        // scattering(_scattering),
+        mic_radius(_mic_radius),
+        mic_radius_sq(mic_radius * mic_radius),
+        sound_speed(_sound_speed),
+        energy_thres(_energy_thres),
+        time_thres(_time_thres),
+        hist_bin_size(_hist_bin_size),
+        max_dist(VAL(room.norm()) + (scalar_t)1.),
+        D(room.size(0)),
+        do_scattering(VAL(_scattering.max()) > (scalar_t)0.),
+        walls(make_walls(_e_absorption, _scattering)) {}
+
+  /**
+   * The main (and only) public entry point of this class. The histograms Tensor
+   * reference is passed along and modified in the subsequent private method
+   * calls. This method spawns num_rays rays in all directions from the source
+   * and calls simul_ray() on each of them.
+   */
+  void compute_histograms(torch::Tensor& histograms) {
+    // TODO: Could probably parallelize call over num_rays? We would need
+    // `num_threads` histograms and then sum-reduce them into a single
+    // histogram.
```

Comment:
can write the description in an affirmative way, something like: `TODO: the for loop can be parallelized over num_rays by ...`